N°5588 - Improve PDF export robustness when AttributeImage dimensions cannot be determined

piRGoif opened the pull request:
Avoid PDF export to crash when having some images GD can't read (in SVG format for example).
In such case we were getting this error log:

```
2022-10-07 01:27:33 | Error   | 1     | Uncaught DivisionByZeroError: Division by zero in /var/www/html/iTop/core/pdfbulkexport.class.inc.php:232
Stack trace:
#0 /var/www/html/iTop/core/htmlbulkexport.class.inc.php(68): PDFBulkExport->GetValue(Object(Person), 'picture')
#1 /var/www/html/iTop/core/tabularbulkexport.class.inc.php(297): HTMLBulkExport->GetSampleData(Object(Person), 'picture')
#2 /var/www/html/iTop/core/htmlbulkexport.class.inc.php(49): TabularBulkExport->GetInteractiveFieldsWidget(Object(AjaxPage), 'interactive_fie...')
#3 /var/www/html/iTop/core/pdfbulkexport.class.inc.php(117): HTMLBulkExport->GetFormPart(Object(AjaxPage), 'interactive_fie...')
#4 /var/www/html/iTop/webservices/export-v2.php(318): PDFBulkExport->GetFormPart(Object(AjaxPage), 'interactive_fie...')
#5 /var/www/html/iTop/webservices/export-v2.php(423): DisplayForm(Object(AjaxPage), 'http://webserve...', 'SELECT `Person`...', NULL, 'pdf')
#6 /var/www/html/iTop/webservices/export-v2.php(689): InteractiveShell('SELECT `Person`...', NULL, 'pdf', '', 'dialog')
#7 {main}
  thrown | IssueLog |||
array (
  'type' => 1,
  'file' => '/var/www/html/iTop/core/pdfbulkexport.class.inc.php',
  'line' => 232,
)
```

Now we won't crash anymore, but we will still get a warning in the log: 
```
2022-12-13 18:53:37 | Warning | 1     | AttributeImage: Cannot read image size | export
array (
  'ObjClass' => 'Person',
  'ObjKey' => '15',
  'ObjFriendlyName' => 'Agatha Christie',
  'AttCode' => 'picture',
)
```

piRGoif commented on core/pdfbulkexport.class.inc.php:
```diff
@@ -190,41 +190,55 @@ public function GetFooter()
-							$iNewWidth = $iWidth * $fScale;
-							$iNewHeight = $iHeight * $fScale;
+							list($iWidth, $iHeight) = utils::GetImageSize($value->GetData());
+							if (($iWidth === 0) && ($iHeight === 0)) {
```
|| instead of && !

piRGoif replied:
Done in b09b665, thanks Molkobain !

Molkobain commented on core/pdfbulkexport.class.inc.php:
```diff
@@ -190,41 +190,55 @@ public function GetFooter()
-							$iNewWidth = $iWidth * $fScale;
-							$iNewHeight = $iHeight * $fScale;
+							list($iWidth, $iHeight) = utils::GetImageSize($value->GetData());
+							if (($iWidth === 0) && ($iHeight === 0)) {
```
```suggestion
							if (($iWidth === 0) || ($iHeight === 0)) {
```

Molkobain commented on core/pdfbulkexport.class.inc.php:
```diff
@@ -190,41 +190,55 @@ public function GetFooter()
-							$iNewHeight = $iHeight * $fScale;
+							list($iWidth, $iHeight) = utils::GetImageSize($value->GetData());
+							if (($iWidth === 0) && ($iHeight === 0)) {
+								// avoid division by zero exception :/
```
```suggestion
								// Avoid division by zero exception (SVGs, corrupted images, ...)
```

Molkobain commented on core/pdfbulkexport.class.inc.php:
```diff
@@ -190,41 +190,55 @@ public function GetFooter()
+								// avoid division by zero exception :/
+								$iNewWidth = $iDefaultMaxWidthPx;
+								$iNewHeight = $iDefaultMaxHeightPx;
+								IssueLog::Warning('AttributeImage : cannot read image size', LogChannels::EXPORT, [
```
```suggestion
								IssueLog::Warning('AttributeImage: Cannot read image size', LogChannels::EXPORT, [
```

Shouldn't this be an error instead of a warning?

piRGoif replied:
Thanks ! About the log level, it won't be an error for svg files for example, and it shouldn't be blocking...
Kept the discussion opened so that it can be discussed on next review !

Molkobain commented:
_As a reminder that was the cause of CI tests failures due to a backup with corrupted images_

Molkobain commented:
_You'll have to rebase this branch as the other PR on PDF export has been merged._

piRGoif commented:
Thanks for pointing that out, I didn't saw there was a conflict...
The modified code was extracted in a new method... No way to rebase easily :(
Too sad, I will certainly miss the review... again :(

piRGoif committed:
Handle returned size 0

piRGoif committed:
Better comment

piRGoif committed:
Fix string

piRGoif committed:
Fix error test : in error if one of width/height is 0, instead of both

piRGoif commented:
Rebase / force push done (wasn't easy :~( )

Molkobain commented:
Accepted during technical review.

Molkobain commented:
Accepted during functional review. Will be part of the upcoming 2.7.9.
@piRGoif will add a reproduction procedure in bug (providing faulty images if possible)

piRGoif commented:
The branch isn't working anymore :(
I'm looking on why (certainly errors made during rebase...)

piRGoif committed:
Fix crashing with "Call to a member function GetData() on null"
The rebase made to get commit 3e18ad59 changes (\PDFBulkExport::GetAttributeImageValue) was not properly done :/
Added the $oObj param in the method signature to be able to keep the info in the log

piRGoif commented:
Fixed the branch, and rebased on support/2.7 (yeah, no conflicts !)

But as I added a param to the \PDFBulkExport::GetAttributeImageValue for logging purpose only I'm setting this PR to be reviewed again on next technical review. What do you think @Molkobain ?

The Combodo ticket was updated with more info (how to reproduce + new log)

Molkobain commented on core/log.class.inc.php:
```diff
@@ -549,6 +549,12 @@ class LogChannels
 
+	/**
+	 * @var string
+	 * @since 2.7.8
```
```suggestion
	 * @since 2.7.9
```

Molkobain commented on core/pdfbulkexport.class.inc.php:
```diff
@@ -275,7 +280,7 @@ public function GetFileExtension()
 	 * @since 2.7.8
 	 */
-	protected function GetAttributeImageValue(AttributeImage $oAttDef, ormDocument $oValue, string $sOutputType)
+	protected function GetAttributeImageValue(DBObject $oObj, AttributeImage $oAttDef, ormDocument $oValue, string $sOutputType)
```
```suggestion
	protected function GetAttributeImageValue(DBObject $oObj, string $sAttCode, string $sOutputType)
```

If we have to pass the whole object, then passing the att. def. and the value makes no sense anymore. We should only pass the object and the att. code IMO.

_Mind to update the PHPDoc and the consumers as well_

piRGoif replied:
Updated the signature with 06259546
Indeed getting again attdef and value shouldn't be that bad as everything is in memory.
Consumers were already updated on last modification, I did it also for this modification.

piRGoif committed:
Update since phpdoc tag

Co-authored-by: Molkobain <lajarige.guillaume@free.fr>

piRGoif committed:
Change \PDFBulkExport::GetAttributeImageValue signature again
Removed attdef and value, to pass only object and attcode
Getting again the attdef and the value won't be costly, they are in memory

Molkobain commented:
Accepted during technical review.

Molkobain commented:
Accepted during functional review.
@piRGoif will ensure that the corresponding ticket is marked to retrofit in 2.7.9 / 3.0.3 / 3.1.0
